pkg/policies/dataplane/response-metrics-registry.go

```diff
@@ -1,10 +1,146 @@
+	"go.uber.org/fx"
+	"go.uber.org/multierr"
+
+	flowcontrolv1 "github.com/fluxninja/aperture/api/gen/proto/go/aperture/flowcontrol/v1"
+	"github.com/fluxninja/aperture/pkg/log"
+	"github.com/fluxninja/aperture/pkg/metrics"
+	"github.com/fluxninja/aperture/pkg/policies/dataplane/iface"
+)
+
 // TODO: Bring metrics to this file
 // 1. per component metrics from actuator, fluxmeter.
 
-// // ResponseMetricsGetAPI is an interface for getting response metrics.
-// type ResponseMetricsGetAPI interface {
-// 	GetFluxMeterHistogram(fluxMeterID string) (prometheus.Histogram, bool)
-// 	GetTokenLatencyHistogram(concurrencyControlID string) (prometheus.Histogram, bool)
-// }
+type responseMetricsRegistry struct {
+	pr                        *prometheus.Registry
+	workloadLatencySummaryVec *prometheus.SummaryVec
+	fluxmeterHistogramVec     *prometheus.HistogramVec
+}
+
+// ProvideResponseMetricsAPI returns API for getting metrics observers.
+func ProvideResponseMetricsAPI(registry *prometheus.Registry, lifecycle fx.Lifecycle) iface.ResponseMetricsAPI {
+	rm := &responseMetricsRegistry{
+		pr: registry,
+	}
+
+	// WORKLOAD LATENCY METRICS
+	rm.workloadLatencySummaryVec = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Name: metrics.WorkloadLatencyMetricName,
+		Help: "Latency summary of workload",
+	}, []string{
+		metrics.PolicyNameLabel,
+		metrics.PolicyHashLabel,
+		metrics.ComponentIndexLabel,
+		metrics.DecisionTypeLabel,
+		metrics.WorkloadIndexLabel,
+	})
+
+	// FLUXMETER METRICS
+	rm.fluxmeterHistogramVec = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: metrics.FluxMeterMetricName,
+		Help: "Latency histogram of fluxmeters",
+	}, []string{
+		metrics.FluxMeterNameLabel,
+		metrics.DecisionTypeLabel,
+		metrics.StatusCodeLabel,
+		metrics.FeatureStatusLabel,
+	})
+
+	lifecycle.Append(fx.Hook{
+		OnStart: func(_ context.Context) error {
+			var errs error
+			err := registry.Register(rm.workloadLatencySummaryVec)
+			if err != nil {
+				if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+					// We're registering this histogram vec from multiple processors
+					// (logs processor and traces processor), so if both processors are
+					// enabled, it's expected that whichever processor is created
+					// second, it will see that the histogram vec was already
+					// registered. Use the existing histogram vec from now on.
+					rm.workloadLatencySummaryVec = are.ExistingCollector.(*prometheus.SummaryVec)
+				} else {
+					errs = multierr.Append(errs, err)
+				}
+			}
+
+			err = registry.Register(rm.fluxmeterHistogramVec)
+			if err != nil {
+				if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+					rm.fluxmeterHistogramVec = are.ExistingCollector.(*prometheus.HistogramVec)
+				} else {
+					errs = multierr.Append(errs, err)
+				}
+			}
+
+			return errs
+		},
+		OnStop: func(_ context.Context) error {
+			unregistered := registry.Unregister(rm.workloadLatencySummaryVec)
```

Comment:
Should write deinit code as reverse of init code as a best practice I.E. De register order should be reverse of Register